### ISSUE TITLE
No span for clippy::large_stack_arrays

### ISSUE THREAD

**Issue by IceTDrinker** [rust-lang/rust-clippy#13774]
### Summary

As asked here: https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/Large.20stack.20array.20without.20code.20span.20clippy.3A.3Alarge_stack_arra/near/485664087

I am posting an issue regarding the clippy::large_stack_arrays lint appearing without a span.

### Reproducer

https://github.com/zama-ai/tfhe-rs/tree/am/chore/clippy-bug (permalink for commit https://github.com/zama-ai/tfhe-rs/commit/e0111f6dd18f9955f271892bb669ed5a30153d44)

Run

RUSTFLAGS="-C target-cpu=native" cargo "+nightly-2024-11-29" clippy --all-targets --features=x86_64-unix,boolean,shortint,integer  -p tfhe -- --no-deps -D warnings

I expected to see this happen:

If I get the large stack array lint, I expect to get the offending span

Instead, this happened:

```
error: allocating a local array larger than 16384 bytes
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#large_stack_arrays
  = note: `-D clippy::large-stack-arrays` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(clippy::large_stack_arrays)]`

error: could not compile `tfhe` (lib test) due to 1 previous error
```

no span is provided

### Version

```text
rustc 1.85.0-nightly (a2545fd6f 2024-11-28)
binary: rustc
commit-hash: a2545fd6fc66b4323f555223a860c451885d1d2b
commit-date: 2024-11-28
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.4
```

**Comment by bobrik** [comment]
We hit this with integration tests having too many entries. I patched clippy locally:

```diff
diff --git a/clippy_lints/src/large_stack_arrays.rs b/clippy_lints/src/large_stack_arrays.rs
index 620e27fa6..3d6fd2d9e 100644
--- a/clippy_lints/src/large_stack_arrays.rs
+++ b/clippy_lints/src/large_stack_arrays.rs
@@ -99,7 +99,7 @@ impl<'tcx> LateLintPass<'tcx> for LargeStackArrays {
                 LARGE_STACK_ARRAYS,
                 expr.span,
                 format!(
-                    "allocating a local array larger than {} bytes",
+                    "allocating a local array larger than {} bytes [element_count = {element_count}, {element_size}]: expr = {expr:#?}",
                     self.maximum_allowed_size
                 ),
                 |diag| {

```

Which gave me this when running the patched binary on a problematic project:

```
warning: allocating a local array larger than 16384 bytes [element_count = 2060, 8]: expr = Expr {
             hir_id: HirId(DefId(0:17633 ~ integration[8ae3]::main).8),
             kind: Array(
                 [
                     Expr {
                         hir_id: HirId(DefId(0:17633 ~ integration[8ae3]::main).9),
                         kind: AddrOf(
                             Ref,
                             Not,
                             Expr {
                                 hir_id: HirId(DefId(0:17633 ~ integration[8ae3]::main).10),
                                 kind: Path(
                                     Resolved(
                                         None,
                                         Path {
                                             span: tests/integration/src/tests/ad_detection/mod.rs:100:7: 100:7 (#704),
                                             res: Def(
                                                 Const,
                                                 DefId(0:2766 ~ integration[8ae3]::tests::ad_detection::test_ad_detector_should_not_run_for_post_request),
                                             ),
                                             segments: [
                                                 PathSegment {
                                                     ident: test_ad_detector_should_not_run_for_post_request#36402,
                                                     hir_id: HirId(DefId(0:17633 ~ integration[8ae3]::main).11),
                                                     res: Def(
                                                         Const,
                                                         DefId(0:2766 ~ integration[8ae3]::tests::ad_detection::test_ad_detector_should_not_run_for_post_request),
                                                     ),
                                                     args: None,
                                                     infer_args: true,
                                                 },
                                             ],
                                         },
                                     ),
                                 ),
                                 span: tests/integration/src/tests/ad_detection/mod.rs:100:7: 120:2 (#704),
                             },
                         ),
                         span: tests/integration/src/tests/ad_detection/mod.rs:100:7: 120:2 (#704),
                     },
                     Expr {
                         hir_id: HirId(DefId(0:17633 ~ integration[8ae3]::main).12),
                         kind: AddrOf(
                             Ref,
                             Not,
                             Expr {
                                 hir_id: HirId(DefId(0:17633 ~ integration[8ae3]::main).13),
                                 kind: Path(
                                     Resolved(
                                         None,
                                         Path {
                                             span: tests/integration/src/tests/ad_detection/mod.rs:81:7: 81:7 (#694),
                                             res: Def(
                                                 Const,
...
```

I'm not entirely sure what's the right way of dealing with this. We can suppress the lint in the test package, but maybe Rust itself should do something differently or Clippy shouldn't complain.

**Comment by matrach** [comment]
I've encountered the same issue and boiled it down to not be affected by our code at all, so there is no span! At least in my case, it is caused by libtest having an array of all test cases (confirmed with cargo expand).

## Steps to reproduce

```shell
cargo init testspam
cd testspam/
mkdir tests
seq 10000 | awk '{print "#[test] fn test_no"$1"(){}"}' | rustfmt > tests/spam.rs
cargo clippy --all-targets -v -- -D clippy::large_stack_arrays 
```

gives:

```
error: allocating a local array larger than 16384 bytes
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#large_stack_arrays
  = note: requested on the command line with `-D clippy::large-stack-arrays`

error: could not compile `testspam` (test "spam") due to 1 previous error

Caused by:
  process didn't exit successfully: `/home/matrach/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/clippy-driver /home/matrach/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc --crate-name spam --edition=2024 tests/spam.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=179 --emit=dep-info,metadata -C embed-bitcode=no -C debuginfo=2 --test --check-cfg 'cfg(docsrs,test)' --check-cfg 'cfg(feature, values())' -C metadata=8c270797cefdc24c -C extra-filename=-4a564de5aecf012e --out-dir /home/matrach/tmp/testspam/target/debug/deps -C incremental=/home/matrach/tmp/testspam/target/debug/incremental -L dependency=/home/matrach/tmp/testspam/target/debug/deps` (exit status: 1)
```

## My versions
```
rustc -Vv
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

@rustbot label +I-false-positive

**Comment by samueltardieu** [comment]
Thanks for the reproducer.
